Add type-checker compatible lazy import mode

pvandyken opened the pull request:
I've been using mkinit for a while to manage imports in a library I'm developing. I'm in the process of switching over to lazy imports using the `lazy_import` library, but doing so messes up all the type annotations. The solution, according to [SPEC1](https://scientific-python.org/specs/spec-0001/#type-checkers), is to use their `lazy_loader.attach_stub` function to read in a `__init__.pyi` interface file. But this would mean writing all the interface files by hand again.

Fortunately, it ended up being quite trivial to reconfigure `mkinit` to write both the `__init__.pyi` and `__init__.py` files as specified by SPEC1. I'm simply writing the vanilla mkinit output into the `.pyi` file, then writing the `attach_stub` boilerplate into `__init__.py`. Within the `__init__.pyi` file, `<AUTOGEN_INIT>` will be respected as usual.

I gated the new mode behind `--lazy_loader_typed`. Some of the files also got touched by my black formatter, so there's currently some formatting changes sprinkled throughout. 

No tests yet, but if you're interested in incorporating this, I can look into writing some.

pvandyken commented on mkinit/__main__.py:
```diff
@@ -174,9 +185,10 @@ def main():
         "with_mods": ns["with_mods"],
         "with_all": ns["with_all"],
-        "relative": ns["relative"],
+        "relative": ns["relative"] or ns["lazy_loader_typed"],
```
Note that the `attach_stub` method requires relative paths, so I always have relative as true for `lazy_loader_typed`

Erotemic replied:
I don't quite understand why `attach_stub` requires relative paths. While I would like it to work with absolute imports (which I tend to prefer because they copy / paste much nicer), it's out of scope for this repo. In any case, I think I would rather either:

1. raise an error if incompatable settings are given
2. fallback to the relative behavior but emit a warning to let the user know we are doing that.

pvandyken replied:
Sounds good, I'll opt for the second option, as it feels wrong to _require_ the use of two args to make the mode work, whereas a warning still keeps everything transparent.

This looks like the relevant issue regarding why relative is needed: https://github.com/scientific-python/lazy_loader/issues/52. It does seem to be arbitrary, but it's not an issue I have time to get into.

pvandyken commented on mkinit/formatting.py:
```diff
@@ -70,12 +71,11 @@ def _insert_autogen_text(modpath, initstr):
     new_lines = lines[:startline] + [initstr_] + lines[endline:]
 
     new_text = "".join(new_lines).rstrip() + "\n"
-    print(new_text)
```
This print never used to be here, and I wasn't sure why everything was being printed when verbose was disabled and `inplace` was set, so I removed it. Let me know if this was a mistake

Erotemic replied:
It's very possible that it's a mistake. I put print statements in for debugging all the time, and I often forget to remove them before committing.

Erotemic commented:
Yes, I am interested in merging this feature. Let's make sure existing tests pass, and new ones for this and corresponding docs would be good.

It may take me a bit to review this, but I will review and help get this feature merged.

FWIW: I prefer triple double quotes for doctests and single quotes  for strings used programatically. Using `pip install xdev` and then `xdev format_quotes <path-to-file>` will modify it to this preference. Normally I don't _really_ care for new code, but it looks like your editor is replacing quotes in places causing the diff to be bigger than it needs to be, so sticking to a consistent style would help speed the review along.

pvandyken commented:
Sounds good, I'll move forward on this when I get a chance, hopefully next week.

pvandyken commented on requirements/tests.txt:
```diff
@@ -33,7 +35,7 @@ coverage>=4.3.4     ; python_version < '3.5' and python_version >= '3.4'    # Py
 coverage>=4.5       ; python_version < '2.7' and python_version >= '2.6'    # Python 2.6
 
-codecov>=2.0.15
+# codecov>=2.0.15
```
This module was not installing on python 3.6. Looking at the [repo](https://github.com/codecov/codecov-python), it seems the package is deprecated and the github actions integration should be used alone. But I didn't explore this extensively, so the CI may not work.

Erotemic replied:
Yes, this package should be removed. I'm working on some maintence updates, rebasing on that (after it is done) should take care of this.

Erotemic replied:
I've merged a branch for version 1.1.0, can you rebase this PR on the latest main?

pvandyken commented:
Okay, with any luck the tests should work now. I didn't end up applying `xdev`, as it changed pretty much the entire repository, so I just went through manually and removed as many of the formatting changes as I could. I couldn't find any tests for the `--lazy_loader` option, so I added that along with `--lazy_loader_typed`. Let me know if you want anything more extensive.

pvandyken commented on mkinit/static_mkinit.py:
```diff
@@ -83,6 +83,7 @@ def autogen_init(
     logger.info(
         "Autogenerating __init__ for modpath_or_name={}".format(modpath_or_name)
     )
+    options = _ensure_options(options)
```
This `_ensure_options` ends up getting called here now in addition to in `formatting._initstr`, but I didn't see any obvious ways to refactor

pvandyken commented on tests/test_with_dummy.py:
```diff
@@ -425,6 +445,38 @@ def test_recursive_lazy_autogen():
+
+    mkinit.autogen_init(
+        pkg_path,
+        options={"lazy_loader": 1, "lazy_loader_typed": True, "relative": True},
```
If relative is not `True`, mkinit runs but attempts at importing the module fail. The CLI currently (silently) sets relative to True whenever `--lazy_loader_typed` is specified, so this shouldn't come up for users, but perhaps you want some extra validation? Not sure where the best place to put it is though

Erotemic commented:
Note: don't worry about failing codecov checks. Test coverage is nice, but I don't have a requirement that it hits a certain percent in this repo (if you know how to disable this check let me know, I've never quite figure that out when working with codecov).

Erotemic commented:
While it would be nice to support relative modules, I think we can consider this PR complete. I just tried the new feature and it's great! I love having the explicit pyi file with the fast lazy loading py file.

Is there a timeline for pushing this feature to pypi? I can make a new release if you plan to need this feature in the pip version, but otherwise I may want to tweak things before a tagged release.

pvandyken commented:
I'm depending on the git reference right now. It's just a dev dependency, so this is fine for the interim. No rush for pypi.

Thanks for merging!

Erotemic commented:
I've made a release, the new feature is now on pypi. Thanks again @pvandyken !

pvandyken commented:
Awesome, thanks so much!
